### ISSUE TITLE
spectrum jitter

### ISSUE THREAD

**Issue by violetmage** [wwmm/pulseeffects#921]
sine wave at frequency that is not multiple of sample rate shows extreme jitter

suggest add averaging previous samples, similar of taa antialiasing

**Comment by wwmm** [comment]
Are you able to get an image or video showing this? The spectrum plugin use in stable releases is from GStreamer. I did not write it. But as the total amount of bins used in the FFT is too high to be drawn efficiently I had to use cubic spline interpolation to get a lower amount of points to draw. Maybe this is adding artifacts.

**Comment by violetmage** [comment]
am wonder if "jitter", is right word

![Peek 2021-03-13 17-57](https://user-images.githubusercontent.com/67824960/111052074-d3fa8880-8425-11eb-9705-ece5f829b8df.gif)

**Comment by wwmm** [comment]
>am wonder if "jitter", is right word

I am not an expert in digital signal processing. So I am not sure... But I would not rule out the possibility that this is an issue in the signal generator.

What happens if you use more than 64 interpolation points in the configuration menu?

**Comment by violetmage** [comment]
exact same

**Comment by throni3git** [comment]
To me, his looks like the spectrum is calculated with a signal that was not windowed with an appropriate window function. This smears the spectrum. Commonly used window functions are hann, hamming or blackman windows for example. [window functions on wikipedia](https://en.wikipedia.org/wiki/Window_function)

Where exactly is the data fed to the spectrum analyzer? I'm new to the code, all i've seen is that `spectrum` from GStreamer is used.

**Comment by wwmm** [comment]
>all i've seen is that spectrum from GStreamer is used

That is right. The current release uses the Spectrum plugin from GStreamer. I haven´t looked at GStremer´s sources in a while but as far as I remember they apply a Hann window by default.

As you can see in this link https://github.com/wwmm/pulseeffects/blob/81da40b4ffd4145b8e5e3309181898ccb7bda698/src/pipeline_base.cpp#L133 we just get the final data sent by the plugin. But as I have to interpolate this data some artifacts could be coming from this. The problem is that in order to have a reasonable resolution at the lowest frequencies I have to make configure the plugin with lots of bins. But that number is too high to be drawn efficiently.

**Comment by throni3git** [comment]
GStreamer uses the Hamming window, which does not completely go down to zero. I have found it [here](https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/-/blob/master/gst/spectrum/gstspectrum.c#L733)

**Comment by wwmm** [comment]
>GStreamer uses the Hamming window, which does not completely go down to zero. I have found it here

If a Hann window solves that problem things will be fine in the gtk4 branch https://github.com/wwmm/pulseeffects/blob/666d73f30df4aadbc8ea42fdbbd66c1ce51aa936/src/spectrum.cpp#L88. I had to write a new spectrum plugin from scratch because native Pipewire filters are being used in the gtk4 branch instead of GStreamer.
